src/utils.py

```diff
@@ -0,0 +1,107 @@
+import logging
+import random
+from typing import Tuple
+
+import requests  # type: ignore
+import tweepy
+from lxml import etree
+
+import constants as c
+import settings as s
+from camera import Camera
+from exceptions import FetchCamerasException, SaveImageException
+from shared import exponential_backoff
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def authenticate_twitter() -> tweepy.API:
+    """Authenticates with the Twitter API and returns a tweepy.API instance."""
+    auth = tweepy.OAuthHandler(c.CONSUMER_KEY, c.CONSUMER_SECRET)
+    auth.set_access_token(c.ACCESS_TOKEN, c.ACCESS_TOKEN_SECRET)
+    return tweepy.API(auth)
+
+
+@exponential_backoff(
+    attempts=s.CAMERA_FETCH_ATTEMPTS,
+    initial_delay=s.INITIAL_RETRY_DELAY,
+    factor=s.RETRY_DELAY_FACTOR,
+    exception_types=(FetchCamerasException, SaveImageException),
+)
+def load_cameras() -> Tuple[str, ...]:
+    """Fetches the camera links and returns them as a tuple."""
+    r = requests.get(c.SITEMAP_URL)
+    r.raise_for_status()
+
+    loc_elements = tuple(link for link in etree.fromstring(r.content).iter("{*}loc"))
+    camera_links = tuple(link.text for link in loc_elements)
+
+    logger.info(f"fetched {len(camera_links)} camera links.")
+    return camera_links
+
+
+def get_random_valid_camera(available_cameras: list[str], camera_constructor) -> Camera:
+    """Returns a random valid Camera object."""
+    while True:
+        random_camera_url = random.choice(available_cameras)
+        camera = camera_constructor(random_camera_url)
+        if not camera.page_content or not camera.stream_url or not camera._url_is_valid():
+            logger.info(f"camera rejected: {camera.id}")
+            continue
+
+        logger.info(f"camera accepted: {camera.id}")
+        return camera
+
+
+def create_tweet_text(camera_info: dict[str, str], flag: str) -> str:
+    """Generates a tweet text based on the camera information and flag."""
+    city = camera_info["city"] if camera_info["city"] != "-" else "Unknown"
+    region = camera_info["region"] if camera_info["region"] != "-" else "Unknown"
+    country = (

```

Comment:
This comprehension feels a bit odd to me because of all the .replace() calls. It looks like it probably works fine but I guess I'd expect to have a dedicated utility function that had a list of pre-defined str values that you knew you'd need to modify, that you could arbitrarily feed stuff into and get something out of.

Nothing wrong with it in practice really but to my mind it feels like it could be made a little more sensible. Any ideas?

Reply:
I went ahead and created a dedicated utility function called `replace_substrings` that takes a string and a dictionary of mappings as input, then replaces the substrings in the string based on the given mappings. I also replaced the comprehension in the create_tweet_text function to use said `replace_substrings` function.

```python
def replace_substrings(string: str, mappings: dict[str, str]) -> str:
    """Replaces substrings in a string based on a dictionary of mappings."""
    for old, new in mappings.items():
        string = string.replace(old, new)
    return string
```